pkg/ipamd/datastore/data_store.go

```diff
@@ -655,67 +563,59 @@ func (ds *DataStore) AssignPodIPv4Address(ipamKey IPAMKey) (ipv4address string,
 
 	ds.log.Debugf("AssignIPv4Address: IP address pool stats: total: %d, assigned %d", ds.total, ds.assigned)
 
-	if eni, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
+	if eni, _, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
 		ds.log.Infof("AssignPodIPv4Address: duplicate pod assign for sandbox %s", ipamKey)
 		return addr.Address, eni.DeviceNumber, nil
 	}
 
 	for _, eni := range ds.eniPool {
-		ds.log.Debugf("Checking for eni %v", eni)
-		if !eni.IsPDEnabled {
-			for _, addr := range eni.IPv4Addresses {
-				if !addr.Assigned() && !addr.inCoolingPeriod() {
-					ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
-					if err := ds.writeBackingStoreUnsafe(); err != nil {
-						ds.log.Warnf("Failed to update backing store: %v", err)
-						// Important! Unwind assignment
-						ds.unassignPodIPv4AddressUnsafe(addr)
-						return "", -1, err
-					}
-
-					return addr.Address, eni.DeviceNumber, nil
+		for _, availableCidr := range eni.AvailableIPv4Cidrs {
+			var addr *AddressInfo
+			var strPrivateIPv4 string
+			var err error
+
+			if ds.isPDEnabled && availableCidr.IsPrefix {
+				strPrivateIPv4, err = ds.getFreeIPv4AddrfromCidr(availableCidr)
+				if err != nil {
+					ds.log.Debugf("Unable to get IP address from prefix: %v", err)
+					//Check in next CIDR
+					continue
+				}
+				ds.log.Debugf("New IP from PD pool- %s", strPrivateIPv4)
+				if availableCidr.IPv4Addresses == nil {
+					availableCidr.IPv4Addresses = make(map[string]*AddressInfo)
```

Comment:
nit: when will this IPv4Addresses ever be nil? seem it never happens from the code i reviewed.

I'm ok to have this safety check, but let's add some comment and debug statement here.
also, this is a sign that we are not encapsulation memory handling  logic good enough.

Reply:
Yes, it is just for safety check, will add a debug